js-packages/@prestojs/viewmodel/src/fields/RelatedViewModelField.ts

```diff
@@ -20,15 +40,148 @@ type RelatedViewModelFieldProps<T extends ViewModelConstructor<any>> = FieldProp
 };
 
 export class UnresolvedRelatedViewModelFieldError<
-    T extends ViewModelConstructor<any> = ViewModelConstructor<any>
+    T extends ViewModelConstructor<any>,
+    FieldValueType extends BaseRelatedViewModelValueType<T>,
+    ParsableType extends RelatedViewModelParsableType<T>
 > extends Error {
-    field: RelatedViewModelField<T>;
-    constructor(field: RelatedViewModelField<T>, message) {
+    field: BaseRelatedViewModelField<T, FieldValueType, ParsableType>;
+    constructor(field: BaseRelatedViewModelField<T, FieldValueType, ParsableType>, message) {
         super(message);
         this.field = field;
     }
 }
 
+/**
+ * We split into RelatedViewModelField (for single records) and ManyRelatedViewModelField (for multiple records)
+ * mainly to making typing easier.
+ *
+ * Use `ManyRelatedViewModelField` if `sourceFieldName` refers to a `ListField` otherwise `RelatedViewModelField`.
+ */
+export abstract class BaseRelatedViewModelField<
+    T extends ViewModelConstructor<any>,
+    FieldValueType extends BaseRelatedViewModelValueType<T>,
+    ParsableType extends RelatedViewModelParsableType<T>
+> extends Field<FieldValueType, ParsableType> {
+    private _loadTo: () => Promise<T> | T;
+    private _resolvedTo: T;
+    private _resolvingTo?: Promise<T>;
+    sourceFieldName: string;
+    sourceField: Field<any>;
+
+    get many(): boolean {
+        return this.sourceField instanceof ListField;
+    }
+
+    constructor(props: RelatedViewModelFieldProps<T, FieldValueType>) {
+        const { to, sourceFieldName, ...fieldProps } = props;
+        super(fieldProps);
+        if (isViewModelClass(to)) {
+            this._resolvedTo = to;
+        } else {
+            this._loadTo = to;
+        }
+        this.sourceFieldName = sourceFieldName;
+        this._isResolvingDeps = false;
+    }
+
+    /**
+     * @private
```

Comment:
Should document what this function does -- as with comment on other review, it will make sense to devs familiar with django internals but not necessarily other frameworks/languages

Reply:
It's documented on the base class